src/mesh_tools.rs

```diff
@@ -28,117 +23,161 @@ use crate::mesh_topology_analysis;
 ///
 /// This method doesn't flip the normals associated with the face vertices, as
 /// there is no unambiguous way to do so automatically.
-
+///
+/// https://gist.github.com/vTurbine/16fbb99225ad4c0ac80b24855dd61a7c
+///
+/// # Warning
+/// As a result, the entire mesh can end up facing inwards (be entirely
+/// reverted). At the moment we have no tools to detect such a case
+/// automatically, so we need to rely on the user to check it and potentially
+/// revert winding of the entire mesh.
+///
+/// # Warning
+/// The results might be unpredictable for non-manifold meshes and moebius-like
+/// topologies.
+///
 // FIXME: Flip also vertex normals if the visual/practical tests prove it's
 // needed
+#[allow(dead_code)]
 pub fn synchronize_mesh_winding(
     geometry: &Geometry,
+    oriented_edges: &[OrientedEdge],
     unoriented_edges: &[UnorientedEdge],
-    edge_to_face_topology: &HashMap<u32, SmallVec<[u32; 8]>>,
+    face_to_oriented_edge_topo: &[SmallVec<[u32; 3]>],
+    unoriented_edge_to_face_topo: &[SmallVec<[u32; 2]>],
```

Comment:
We need `UnorientedEdge`s to find face's neighbors and identify the edge they share. It's also the edge that should be reverted in each of the faces. We might get rid of `oriented_edges` here somehow though: for example there can be a method of the `TriangleFace` that would take an `UnorientedEdge` as an argument and return a respective `OrientedEdge`. It might be actually quite efficient but it's a different approach.